xbmc/music/MusicDatabase.cpp

```diff
@@ -5760,30 +6022,59 @@ bool CMusicDatabase::GetFilter(CDbUrl &musicUrl, Filter &filter, SortDescription
+        strSQL += PrepareSQL("(SELECT song_artist.idArtist FROM song_artist" // All artists linked to extra genres
+        " JOIN song_genre ON song_artist.idSong = song_genre.idSong"
+        " WHERE song_genre.idGenre = %i AND song_artist.idRole = %i )", idGenre, idRole);
+
+      if (idRole < -1) //All artists contributing to songs, all roles.
+        strSQL = PrepareSQL("(artistview.idArtist IN (SELECT song_artist.idArtist FROM song_artist"
+        " JOIN song_genre ON song_artist.idSong = song_genre.idSong"
+        " WHERE song_genre.idGenre = %i)"
+        " OR artistview.idArtist IN"
+        " (SELECT album_artist.idArtist FROM album_artist" // All album artists linked to extra genres
+        " JOIN album_genre ON album_artist.idAlbum = album_genre.idAlbum"
+        " WHERE album_genre.idGenre = %i)",
+        idGenre, idGenre);
     }
     else
-    {
-      if (!albumArtistsOnly)  // show all artists in this case (ie those linked to a song)
-        strSQL += "(SELECT song_artist.idArtist FROM song_artist)"
-                  " OR artistview.idArtist IN ";
-
-      // and always show any artists linked to an album (may be different from above due to album artist tag)
-      strSQL += "(SELECT album_artist.idArtist FROM album_artist)"; // Includes compliation albums hence "Various artists"
+    { // Artists can be only album artists, so for all artists (with linked albums or songs) 
+      // we need to check both album_artist and song_artist tables.
+      // Role is determined from song_artist table, so even if looking for album artists only
+      // we can check those that have a specific role e.g. which album artist is a composer
+      // (but not neccessarily of songs in that album), from entries in the song_artist table.
+      // Role < -1 is used to indicate that all roles are wanted.
+      // When not album artists only and a specific role wanted then only the song_artist table is checked.
+      // When album artists only and role = 1 (an "artist") then only the album_artist table is checked.
+      if (albumArtistsOnly || idRole == 1)  //Album artists only or no specific role, check album_artist table
+        strSQL += "(SELECT album_artist.idArtist FROM album_artist )"; // Includes compliation albums hence "Various artists"
+      if (albumArtistsOnly && idRole > 1)  //Album artists only with role, check AND in song_artist
+        strSQL += " AND artistview.idArtist IN ";
+      if (!albumArtistsOnly && idRole == 1) //Song and album artists no specific role, check OR in song_artist
+        strSQL += " OR artistview.idArtist IN ";
+      if (!albumArtistsOnly || idRole > 1)  //Song and album artists or album artists with specific role, check song_artist
+        strSQL += PrepareSQL("(SELECT song_artist.idArtist FROM song_artist WHERE song_artist.idRole = %i)", idRole);
+
+      if (idRole < -1) //All artists contributing to songs, all roles.
```

Comment:
should it really be < -1? I haven't seen it set to lower than -1 so maybe <= or == ?